### PR TITLE
Speed up sum by using reasonable read buffer sizes.

### PR DESCRIPTION
**Pull request by resistor**
Use a 4K read buffer for each of the checksum functions, which seems
reasonable. This improves the performance of BSD checksums on
odyssey1024.txt from 399ms to 325ms on my laptop, and of SysV
checksums from 242ms to 67ms.

### CODE REVIEW

**Comment by sylvestre** [comment]
can you please document your benchmark with hyperfine in src/uu/sum/BENCHMARKING.md ?
thanks

**Comment by sylvestre** [comment]
it broke some tests btw:

```

---- test_sum::test_sysv_stdin stdout ----
current_directory_resolved: 
open: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.tmpz1HceO/lorem_ipsum.txt
run: /Users/runner/work/coreutils/coreutils/target/debug/coreutils sum -s
open: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.tmpz1HceO/sysv_stdin.expected
---- test_sum::test_sysv_stdin stderr ----
thread 'main' panicked at 'assertion failed: `(left == right)`
Diff< left / right > :
 [
     54,
     57,
     56,
     53,
     32,
<    49,
>    50,
     10,
 ]
', tests/common/util.rs:261:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by resistor** [comment]
> it broke some tests btw:
> 
> ```
> 
> ---- test_sum::test_sysv_stdin stdout ----
> current_directory_resolved: 
> open: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.tmpz1HceO/lorem_ipsum.txt
> run: /Users/runner/work/coreutils/coreutils/target/debug/coreutils sum -s
> open: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.tmpz1HceO/sysv_stdin.expected
> ---- test_sum::test_sysv_stdin stderr ----
> thread 'main' panicked at 'assertion failed: `(left == right)`
> Diff< left / right > :
>  [
>      54,
>      57,
>      56,
>      53,
>      32,
> <    49,
> >    50,
>      10,
>  ]
> ', tests/common/util.rs:261:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

Fixed, thanks.

**Comment by resistor** [comment]
> can you please document your benchmark with hyperfine in src/uu/sum/BENCHMARKING.md ?
> thanks

Done

File: src/uu/sum/src/sum.rs [hunk 1]
```diff
@@ -63,6 +69,7 @@ fn sysv_sum(mut reader: Box<dyn Read>) -> (usize, u16) {
     ret = (ret & 0xffff) + (ret >> 16);
     ret = (ret & 0xffff) + (ret >> 16);
 
+    let blocks_read = div_ceil(bytes_read, 512);
```
Comment (sylvestre) [inline src/uu/sum/src/sum.rs:72]: please a comment explaining the why here, thanks

Comment (resistor) [reply src/uu/sum/src/sum.rs:74]: Done.

File: src/uu/sum/src/sum.rs [hunk 2]
```diff
@@ -23,14 +23,19 @@ static USAGE: &str = "{} [OPTION]... [FILE]...";
 static SUMMARY: &str = "Checksum and count the blocks in a file.\n\
                         With no FILE, or when  FILE is -, read standard input.";
 
+// This can be replaced with usize::div_ceil once it is stabilized
+fn div_ceil(a: usize, b: usize) -> usize {
+    (a + b - 1) / b
```
Comment (anastygnome) [inline src/uu/sum/src/sum.rs:28]: You should use 
```rust
        pub const fn div_ceil(a : usize, b: usize) -> usize {
            let d = a / b;
            let r= a % b;
            if r > 0 { // round up if not a multiple
                d + 1
            } else {
                d
            }
        }
```
to allow the compuler to use both the division and remainder. with only one asm instruction
(don't forget to use cargo fmt)

File: src/uu/sum/BENCHMARKING.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+and then you can time how it long it takes to checksum the file by running
+
+```shell
+$ /usr/bin/time ./target/release/sum wikidatawiki-20211001-pages-logging.xml
+```
+
+For more systematic measurements that include warm-ups, repetitions and comparisons, [Hyperfine](https://github.com/sharkdp/hyperfine) can be helpful. For example, to compare this implementation to the one provided by your distribution run
+
+```shell
+$ hyperfine "./target/release/sum wikidatawiki-20211001-pages-logging.xml" "/usr/bin/sum wikidatawiki-20211001-pages-logging.xml"
```
Comment (anastygnome) [inline src/uu/sum/BENCHMARKING.md:22]: remove the /usr/bin from path for portability

**Review by anastygnome — CHANGES_REQUESTED** [review]

File: src/uu/sum/src/sum.rs (hunk 2 above)
Comment (resistor) [reply src/uu/sum/src/sum.rs:28]: When `b` is a power of two (as it is in this file), the original version versions better code. See https://godbolt.org/z/85fnajWhc

File: src/uu/sum/BENCHMARKING.md (hunk 3 above)
Comment (resistor) [reply src/uu/sum/BENCHMARKING.md]: Done. Note that this was taken verbatim from `BENCHMARKING.md` for `tac`.

File: src/uu/sum/src/sum.rs (hunk 1 above)
Comment (resistor) [reply src/uu/sum/src/sum.rs:74]: Actually, the original version is better even for non-power-of-two values of `b` as well, though not as substantially: https://godbolt.org/z/njP8qvqs8

**Comment by sylvestre** [comment]
On my SSD system:
```
  './target/release/coreutils sum wikidatawiki-latest-pages-logging.xml' ran
    1.42 ± 0.09 times faster than './target/release/coreutils.before sum wikidatawiki-latest-pages-logging.xml'
    2.07 ± 0.03 times faster than 'sum wikidatawiki-latest-pages-logging.xml'
```

bravo for the improvement :)

File: src/uu/sum/src/sum.rs (hunk 2 above)
Comment (anastygnome) [reply src/uu/sum/src/sum.rs:28]: @resistor yes, indeed the compiler looks smart enough in that case. My apologies.

Why not declairng the fonction constant though ?

Comment (resistor) [reply src/uu/sum/src/sum.rs:30]: Done

**Pull request merged**